N8N-3727 N8N-3725 pin data empty state and size check

alexgrozav opened the pull request:
- Various UX improvements.
- Empty state
- Size check

alexgrozav committed:
feat: Size check. Various design and ux improvements.

ivov commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -93,10 +94,11 @@
 					type="tertiary"
-					:circle="false"
-					@click="enterEditMode"
+					@click="enterEditMode()"
```
```suggestion
					@click="enterEditMode"
```

alexgrozav replied:
First argument would be the click event object if we don't provide the `()`. Did this because `enterEditMode` is a method with arguments.

ivov commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -286,7 +297,7 @@ import {
 	IDataObject,
 	INodeExecutionData,
-	INodeTypeDescription,
+	INodeTypeDescription, IRun,
```
Nitpick
```suggestion
	INodeTypeDescription,
	IRun,
```

ivov commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -303,7 +314,7 @@ import {
 import {
 	MAX_DISPLAY_DATA_SIZE,
-	MAX_DISPLAY_ITEMS_AUTO_ALL,
+	MAX_DISPLAY_ITEMS_AUTO_ALL, MAX_WORKFLOW_PINNED_DATA_SIZE,
```
Nitpick
```suggestion
	MAX_DISPLAY_ITEMS_AUTO_ALL,
	MAX_WORKFLOW_PINNED_DATA_SIZE,
```

ivov commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -433,7 +445,7 @@ export default mixins(
 			},
 			hasNodeRun(): boolean {
-				return Boolean(!this.isExecuting && this.node && this.workflowRunData && this.workflowRunData.hasOwnProperty(this.node.name));
+				return Boolean(!this.isExecuting && this.node && (this.workflowRunData && this.workflowRunData.hasOwnProperty(this.node.name) || this.hasPinData));
```
`hasPinData` is independent of whether node has run. I would hoist `this.hasPinData` alongside `hasNodeRun` instead of nesting it inside.

ivov commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -578,14 +603,50 @@ export default mixins(
+					}
+				}
+			},
+			verifyPinnedDataSize(data: IDataObject[]): boolean {
+				try {
+					if (this.$store.getters['pinDataSize'] + stringSizeInBytes(JSON.stringify(data)) > MAX_WORKFLOW_PINNED_DATA_SIZE) {
+						throw new Error(this.$locale.baseText('ndv.pinData.error.tooLarge.description'));
+					}
+
+					return true;
+				} catch (error) {
+					this.$showError(error, this.$locale.baseText('ndv.pinData.error.tooLarge.title'));
+					return false;
```
1. Looks like whenever you call `stringSizeInBytes` you need to stringify first. Maybe we should stringify inside the function.
2. Instead of a try-catch, I think this would be more readable with an if check.
3. Since it returns a boolean, naming: `isValidPinDataSize`?

alexgrozav replied:
- Changed to if check instead of try-catch. 
- Renamed to `isValidPinDataSize`. 
- Now accepting both strings and objects, and stringifying values if they're not of string type

ivov commented on packages/editor-ui/src/components/helpers.ts:
```diff
@@ -67,3 +67,7 @@ export function isString(value: unknown): value is string {
 	return typeof value === 'number';
 }
+
+export function stringSizeInBytes(input: string): number {
+	return encodeURI(input).split(/%(?:u[0-9A-F]{2})?[0-9A-F]{2}|./).length - 1;
+}
```
Looks a bit cryptic.

alexgrozav replied:
UTF-16 strings can have characters like `%uXX` and `%uXXXX`. 

Essentially this has a lookahead for `%u` + a pair of optional hexadecimal characters `([0-9A-F]{2})?` + another `[0-9A-F]{2}` pair of hexadecimal values, or anything else `|.`.

ivov commented on packages/editor-ui/src/constants.ts:
```diff
@@ -1,3 +1,5 @@
+export const MAX_WORKFLOW_SIZE = 16777216;
+export const MAX_WORKFLOW_PINNED_DATA_SIZE = 12582912;
```
Let's add the unit in the const name or as a comment.

alexgrozav replied:
Done.

ivov commented on packages/editor-ui/src/plugins/i18n/locales/en.json:
```diff
@@ -351,8 +353,16 @@
+	"ndv.pinData.unpinOnExecute.cancel": "Cancel",
+	"ndv.pinData.unpinOnExecute.confirm": "Unpin and execute",
+	"ndv.pinData.error.tooLarge.title": "Output data is too large to pin",
+	"ndv.pinData.error.tooLarge.description": "You can pin at most 12MB of output per workflow.",
```
```suggestion
	"ndv.pinData.error.tooLarge.description": "You can pin at most 12 MB of output per workflow.",
```

ivov replied:
We're not doing this e.g. in `ndv.output.tooMuchData.message`

ivov commented on packages/editor-ui/src/plugins/i18n/locales/en.json:
```diff
@@ -650,6 +660,7 @@
 	"runData.table": "Table",
 	"runData.editor.save": "Save",
 	"runData.editor.cancel": "Cancel",
+	"runData.editor.copyDataInfo": "You can copy data from previous executions and paste them above",
```
```suggestion
	"runData.editor.copyDataInfo": "You can copy data from previous executions and paste it above",
```

alexgrozav replied:
Done.

ivov reviewed (commented):
Nice work!

ivov commented:
Findings:
1. If the user toggles on `JSON`, inserts data and clicks `Save`, the output pane toggles back to `Table`.
2. When pushing the center pane to the right, the last word in the footer overflows.
3. When pushing the center to the right until the footer text grows to three lines, the spacing between pindata editor and top buttons is lost.
4. The `JSON/Table` toggle seems to have a bit more bottom padding than top padding. According to the design it should be centered in its container.
5. Pin button should not be showing on input pane.
6. When the pindata editor is open and the user clicks on `Execute node`, I would expect the pindata editor to discard changes and autoclose. Maybe a question for Product.
7. The pindata editor seems to be misvalidating JSON, e.g. `{ hello: true }` triggers `; expected`

ivov committed:
:zap: Add transformer

ivov committed:
:zap: Hoist pin data

ivov committed:
:zap: Adjust endpoints for hoisted pin data

ivov committed:
:blue_book: Expand interface

ivov committed:
:bug: Fix stray array

ivov committed:
:shirt: Fix build

ivov committed:
:shirt: Add lint exception

ivov committed:
:shirt: Fix header

alexgrozav committed:
Merge pull request #3570 from n8n-io/n8n-3927-hoist-data-on-frontend

Hoist pin data on frontend

alexgrozav committed:
Merge branch 'n8n-3722-add-data-pinning-functionality-to-node' into n8n-3727-pin-data-empty-state

alexgrozav commented on packages/editor-ui/src/components/RunData.vue:
```diff
@@ -26,9 +26,9 @@
 								{{ $locale.baseText('ndv.pinData.pin.description') }}
 							</p>
-							<n8n-link to="https://google.com" size="s">
+							<n8n-link to="https://google.com" size="small">
```
Done.

alexgrozav commented:
> Findings:
> 
> 1. If the user toggles on `JSON`, inserts data and clicks `Save`, the output pane toggles back to `Table`.
> 2. When pushing the center pane to the right, the last word in the footer overflows.
> 3. When pushing the center pane to the right until the footer text grows to three lines, the spacing between pindata editor and top buttons is lost.
> 4. The `JSON/Table` toggle seems to have a bit more bottom padding than top padding. According to the design it should be centered in its container.
> 5. Pin button should not be showing on input pane.
> 6. When the pindata editor is open and the user clicks on `Execute node`, I would expect the pindata editor to discard changes and autoclose. Maybe a question for Product.
> 7. The pindata editor seems to be misvalidating JSON, e.g. `{ hello: true }` triggers `; expected`

1. JSON / Table toggle is no longer visible while editing
2. I suppose this is intended. Should we have a minimum width? 
3. There's something weird going on with Monaco's resize listener. It seems to be lagging behind for some reason. Need to investigate.
4. Fixed in master.
5. Done.
6. Added prompt. Done.
7. Validation working now! :)

ivov commented:
Thanks for addressing everything!

2. From what I remember, the last word overflowed pretty badly, so it cannot have been intended, unless other changes since made it not look too bad now. In any case, if needed truncation would be nice. Is rather an edge case.
